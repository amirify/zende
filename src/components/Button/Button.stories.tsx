import type { Meta, StoryObj } from "@storybook/react";
import { StoryContainer } from "../StoryContainer";
import { useState } from "react";
import { Button } from "./Button";
import { ButtonProps, ButtonVariant } from "./Button.types";
import { LoadingIcon } from "../Icons/LoadingIcon";

const meta: Meta<typeof Button> = {
  component: Button,
  title: "Button",
  tags: ["autodocs"],
  includeStories: /^[A-Z]/,
  argTypes: {
    onClick: { action: "clicked" },
  },
  decorators: [
    (Story) => {
      return (
        <StoryContainer>
          <Story />
        </StoryContainer>
      );
    },
  ],
};

export default meta;
type Story = StoryObj<ButtonProps>;

// Generic helper function to generate variant-based button stories
const createVariantStory = <K extends keyof ButtonProps>(
  variantProp: K,
  value: ButtonProps[K],
  label: string,
): Story => ({
  args: { [variantProp]: value },
  render: (args) => (
    <>
      <Button {...args} variant="solid">
        {label}
      </Button>
      <Button {...args} variant="outlined">
        {label}
      </Button>
      <Button {...args} variant="text">
        {label}
      </Button>
    </>
  ),
});

// Default story
export const Default: Story = {
  args: {
    children: "Default Button",
  },
};

// Variants
export const SolidVariant: Story = {
  args: {
    children: "Solid Button",
    variant: "solid",
  },
};

export const OutlinedVariant: Story = {
  args: {
    children: "Outlined Button",
    variant: "outlined",
  },
};

export const TextVariant: Story = {
  args: {
    children: "Text Button",
    variant: "text",
  },
};

// Sizes
export const SmallSize = createVariantStory("size", "small", "Small Button");
export const MediumSize = createVariantStory("size", "medium", "Medium Button");
export const LargeSize = createVariantStory("size", "large", "Large Button");

// Radii
export const NoRadius = createVariantStory("radius", "none", "No Radius");
export const SmallRadius = createVariantStory("radius", "sm", "Small Radius");
export const MediumRadius = createVariantStory("radius", "md", "Medium Radius");
export const LargeRadius = createVariantStory("radius", "lg", "Large Radius");
export const PillRadius = createVariantStory("radius", "pill", "Pill Radius");
export const CircleRadius = createVariantStory("radius", "circle", "Circle");

// Block
export const BlockButton = createVariantStory("block", true, "Block Button");

// Disabled
export const DisabledButton = createVariantStory(
  "disabled",
  true,
  "Disabled Button",
);

// Icon position
export const IconAtStart: Story = {
  args: {
    children: "Icon Start",
    icon: <LoadingIcon margin="right" />,
    iconPosition: "start",
  },
  render: (args) => (
    <>
      {(["solid", "outlined", "text"] as ButtonVariant[]).map((variant) => (
        <Button key={variant} {...args} variant={variant}>
          {args.children}
        </Button>
      ))}
    </>
  ),
};

export const IconAtEnd: Story = {
  args: {
    children: "Icon End",
    icon: <LoadingIcon margin="left" />,
    iconPosition: "end",
  },
  render: (args) => (
    <>
      {(["solid", "outlined", "text"] as ButtonVariant[]).map((variant) => (
        <Button key={variant} {...args} variant={variant}>
          {args.children}
        </Button>
      ))}
    </>
  ),
};

export const LoadingState: Story = (args: ButtonProps) => {
  const [showLoading, setShowLoading] = useState(args.showLoading);

  return (
    <>
      {(["solid", "outlined", "text"] as ButtonVariant[]).map((variant) => (
        <Button
          {...args}
          key={variant}
          showLoading={showLoading}
          variant={variant}
          onClick={() => setShowLoading(!showLoading)}
        >
          {args.children}
        </Button>
      ))}
    </>
  );
};

// Default args for the story
LoadingState.args = {
  children: "Toggle Loading",
  showLoading: false,
};
