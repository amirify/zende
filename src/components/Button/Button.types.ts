import { ReactNode, CSSProperties } from 'react';

export type ButtonVariant = 'solid' | 'outlined' | 'text';
export type ButtonSize = 'small' | 'medium' | 'large';
export type ButtonRadius = 'none' | 'sm' | 'md' | 'lg' | 'pill' | 'circle';
export type ButtonIconPosition = 'start' | 'end';

export interface ButtonProps {
    children: ReactNode;
    disabled?: boolean;
    onClick?: () => void;
    variant?: ButtonVariant;            // default: 'solid'
    size?: ButtonSize;                  // default: 'medium'
    radius?: ButtonRadius;              // default: 'none'
    icon?: ReactNode;
    iconPosition?: ButtonIconPosition;  // default: 'start'
    loadingIcon?: ReactNode;
    showLoading?: boolean;              // If true, shows loadingIcon instead of children
    block?: boolean;                    // If true, the button takes full width of its container
    className?: string;                 // Additional custom class names for the button
    styles?: CSSProperties;             // Inline styles to apply to the button
}