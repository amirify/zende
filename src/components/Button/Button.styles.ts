import { css, SerializedStyles, keyframes } from '@emotion/react';
import { Theme } from '../../theme/theme.types';
import { ButtonVariant, ButtonSize, ButtonRadius } from './Button.types';

export const computeButtonStyles = (theme: Theme, variant: ButtonVariant, size: ButtonSize, radius: ButtonRadius, block: boolean): SerializedStyles[] => {

  const rotateBackground = keyframes`
    from {
      transform: translate(-50%, -50%) rotate(0);
    }
    to {
      transform: translate(-50%, -50%) rotate(360deg);
    }
 `;

  const pulseShadow = keyframes`
    0% {
      box-shadow: 0 0 0 1px ${theme.colors.primary};
      color: ${theme.colors.primary};
    }
    50% {
      box-shadow: 0 0 0 3px ${theme.colors.secondary};
      color: ${theme.colors.secondary};
    }
    100% {
      box-shadow: 0 0 0 1px ${theme.colors.primary};
      color: ${theme.colors.primary};
    }
 `;


  const buttonVariantMappings: Record<ButtonVariant, SerializedStyles> = {
    solid: css({
      color: theme.colors.onPrimary,
      boxShadow: theme.shadows.lg,
      '&::before': {
        content: '""',
        display: 'block',
        width: '120%',
        height: '0',
        paddingTop: '120%',
        position: 'absolute',
        top: '50%',
        left: '50%',
        background: `linear-gradient(${theme.colors.primary}, ${theme.colors.secondary})`,
        animation: `${rotateBackground} 3s linear infinite`,
        zIndex: -2,
      },
      '&::after': {
        content: '""',
        display: 'block',
        width: '250%',
        height: '100%',
        position: 'absolute',
        top: 0,
        left: 0,
        transform: 'translate(-100%, 0) rotate(-45deg)',
        background: 'rgba(255, 255, 255, 0.3)',
        zIndex: -1
      },
      '&:hover::after': {
        transition: 'all 1000ms ease',
        transform: 'translate(100%, 0) rotate(-45deg)'
      },
      '&:active': {
        boxShadow: theme.shadows.md
      },
      '&:disabled': {
        background: theme.colors.neutral,
        color: theme.colors.onNeutral,
        boxShadow: 'none',
        '&::before, &::after': {
          display: 'none'
        }
      }
    }),
    outlined: css({
      animation: `${pulseShadow} 3s infinite ease-in-out`,
      '&:disabled': {
        animation: 'none',
        color: theme.colors.neutral,
        border: `1px solid ${theme.colors.neutral}`
      }
    }),
    text: css({
      color: theme.colors.primary,
      '&::before': {
        content: '""',
        display: 'block',
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        transform: 'scaleY(0)',
        background: `linear-gradient(90deg, ${theme.colors.primary}, ${theme.colors.secondary})`,
        transition: 'all 0.3s ease',
        opacity: 0.3,
        zIndex: -1,
      },
      '&:hover::before': {
        transform: 'scaleY(1)',
      },
      '&:disabled': {
        color: theme.colors.neutral,
        '&::before': {
          display: 'none'
        }
      }
    })
  };

  const buttonSizeMappings: Record<ButtonSize, SerializedStyles> = {
    small: css({
      padding: `${theme.spacing.xs} ${theme.spacing.sm}`,
      fontSize: theme.typography.fontSizes.sm
    }),
    medium: css({
      padding: `${theme.spacing.sm} ${theme.spacing.md}`,
      fontSize: theme.typography.fontSizes.md
    }),
    large: css({
      padding: `${theme.spacing.md} ${theme.spacing.lg}`,
      fontSize: theme.typography.fontSizes.md
    })
  };

  const circleButtonSizeMappings: Record<ButtonSize, SerializedStyles> = {
    small: css({
      width: '2.5rem',
      height: '2.5rem',
      fontSize: theme.typography.fontSizes.xs
    }),
    medium: css({
      width: '3.5rem',
      height: '3.5rem',
      fontSize: theme.typography.fontSizes.md
    }),
    large: css({
      width: '4.5rem',
      height: '4.5rem',
      fontSize: theme.typography.fontSizes.lg
    }),
  };


  const buttonRadiusMappings: Record<ButtonRadius, SerializedStyles> = {
    none: css({ borderRadius: '0' }),
    sm: css({ borderRadius: theme.radii.sm }),
    md: css({ borderRadius: theme.radii.md }),
    lg: css({ borderRadius: theme.radii.lg }),
    xl: css({ borderRadius: theme.radii.xl }),
    pill: css({ borderRadius: theme.radii.pill }),
    circle: css({ borderRadius: theme.radii.circle })
  };

  const buttonBaseClass: SerializedStyles = css({
    position: 'relative',
    overflow: 'hidden',
    border: 'none',
    outline: 'none',
    background: 'none',
    cursor: 'pointer',
    transition: '0.3s all ease',
    zIndex: 0,
    '&:hover': {
      transform: 'scale(1.05) translateZ(0)'
    },
    '&:active': {
      transform: 'scale(1) translateZ(0)'
    },
    '&:disabled': {
      transform: 'none',
      cursor: 'not-allowed',
    }
  });

  const blockButtonClass: SerializedStyles | null = block && radius !== 'circle' ? css({ display: 'block', width: '100%' }) : null;
  const buttonVariantClass: SerializedStyles = buttonVariantMappings[variant];
  const buttonSizeClass: SerializedStyles = radius === 'circle' ? circleButtonSizeMappings[size] : buttonSizeMappings[size];
  const buttonRadiusClass: SerializedStyles = buttonRadiusMappings[radius];

  const buttonClasses: SerializedStyles[] = [buttonBaseClass, buttonVariantClass, buttonSizeClass, buttonRadiusClass];

  if (blockButtonClass) {
    buttonClasses.push(blockButtonClass);
  }

  return buttonClasses;
}