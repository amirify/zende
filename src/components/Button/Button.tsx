import { useTheme, SerializedStyles } from '@emotion/react';
import { Theme } from '../../theme/theme.types';
import { lightTheme } from '../../theme/themes';
import { ButtonProps } from './Button.types';
import { computeButtonStyles } from './Button.styles';
import { LoadingIcon } from '../Icons/LoadingIcon';

export const Button: React.FC<ButtonProps> = ({
    children,
    disabled = false,
    onClick,
    variant = 'solid',
    size = 'medium',
    radius = 'none',
    icon,
    iconPosition = 'start',
    loadingIcon = <LoadingIcon />,
    showLoading = false,
    block = false,
    className,
    styles
}) => {
    // Get theme from Emotion's useTheme hook
    const theme: Theme = useTheme();

    // Fallback to default theme if useTheme returns empty object or undefined
    const activeTheme: Theme = Object.keys(theme).length ? theme : lightTheme;

    // Compute button styles based on theme and provided props
    const buttonStyles: SerializedStyles[] = computeButtonStyles(activeTheme, variant, size, radius, block);

    // Variable to handle icon and children rendering based on iconPosition
    const content = icon && iconPosition === 'end' ? (
        <>
            {children} {icon}
        </>
    ) : (
        <>
            {icon} {children}
        </>
    );

    return (
        <button
            onClick={onClick}
            disabled={disabled}
            css={buttonStyles}
            className={className}
            style={styles}
        >
            <span style={{ visibility: showLoading ? 'hidden' : 'visible' }}
                css={{
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center'
                }}>
                {content}
            </span>
            <span style={{ visibility: showLoading ? 'visible' : 'hidden' }}
                css={{
                    position: 'absolute',
                    top: 0,
                    left: 0,
                    width: '100%',
                    height: '100%',
                    display: 'flex',
                    justifyContent: 'center',
                    alignItems: 'center'
                }}>
                {loadingIcon}
            </span>
        </button>
    );
};