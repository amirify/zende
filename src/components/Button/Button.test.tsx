import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import { Button } from './Button';

// Rendering the Button with Default Properties
test('renders button with default properties', () => {
	render(<Button>Default Button</Button>);

	const button = screen.getByRole('button', { name: /default button/i });
	expect(button).toBeInTheDocument();
	expect(button).toHaveTextContent('Default Button');
});

// Handling disabled State
test('renders disabled button and prevents clicks', async () => {
	const handleClick = jest.fn();
	render(<Button disabled onClick={handleClick}>Disabled Button</Button>);
	
	const button = screen.getByRole('button', { name: /disabled button/i });
	expect(button).toBeDisabled();
	
	await userEvent.click(button);
	expect(handleClick).not.toHaveBeenCalled();
});

// Rendering with Different Variants
test('renders button with variant: outlined', () => {
	render(<Button variant="outlined">Outlined Button</Button>);
	
	const button = screen.getByRole('button', { name: /outlined button/i });
	expect(button).toBeInTheDocument();
});

// Rendering with Different Sizes
test('renders button with size: large', () => {
	render(<Button size="large">Large Button</Button>);
	
	const button = screen.getByRole('button', { name: /large button/i });
	expect(button).toBeInTheDocument();
});

// Displaying an Icon in Different Positions
test('renders button with an icon at the start', () => {
	render(<Button icon={<span data-testid="icon">🔍</span>} iconPosition="start">Icon Button</Button>);
	
	const button = screen.getByRole('button', { name: /icon button/i });
	const icon = screen.getByTestId('icon');
	expect(button).toContainElement(icon);
});

// Displaying Loading Icon Instead of Label When showLoading Prop Is True
test('toggles visibility between loading icon and button text based on showLoading', () => {
	render(<Button showLoading loadingIcon={<span data-testid="loading-icon">⏳</span>}>Button Text</Button>);

	// Check that the loading icon is visible
	const loadingIcon = screen.getByTestId('loading-icon');
	expect(loadingIcon).toBeVisible();
	
	// Check that the button text is not visible
	const buttonText = screen.getByText('Button Text');
	expect(buttonText).not.toBeVisible();
});

// Applying block Style
test('renders button with full width when block is true', () => {
	render(<Button block>Full Width Button</Button>);
	
	const button = screen.getByRole('button', { name: /full width button/i });
	expect(button).toBeInTheDocument();
	// We assume a CSS class or style would make the button full-width, which can be checked visually or with snapshot testing
});

// Click Event Handling
test('calls onClick handler when clicked', async () => {
	const handleClick = jest.fn();
	render(<Button onClick={handleClick}>Clickable Button</Button>);
	
	const button = screen.getByRole('button', { name: /clickable button/i });
	await userEvent.click(button);
	expect(handleClick).toHaveBeenCalledTimes(1);
});

// Snapshot Test
test('Matches snapshot', () => {
  const { asFragment } = render(<Button>Test Button</Button>);
  expect(asFragment()).toMatchSnapshot();
});