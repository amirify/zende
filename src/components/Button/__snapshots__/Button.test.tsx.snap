// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="css-l33wsp-buttonBaseClass-solid-medium-none"
  >
    <span
      class="css-1ezqju9-Button"
      style="visibility: visible;"
    >
       Test Button
    </span>
    <span
      class="css-61oj3j-Button"
      style="visibility: hidden;"
    >
      <svg
        fill="currentColor"
        height="24px"
        viewBox="0 0 128 128"
        width="24px"
        xml:space="preserve"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g>
          <circle
            cx="16"
            cy="64"
            r="16"
          />
          <circle
            cx="16"
            cy="64"
            r="14.344"
            transform="rotate(45 64 64)"
          />
          <circle
            cx="16"
            cy="64"
            r="12.531"
            transform="rotate(90 64 64)"
          />
          <circle
            cx="16"
            cy="64"
            r="10.75"
            transform="rotate(135 64 64)"
          />
          <circle
            cx="16"
            cy="64"
            r="10.063"
            transform="rotate(180 64 64)"
          />
          <circle
            cx="16"
            cy="64"
            r="8.063"
            transform="rotate(225 64 64)"
          />
          <circle
            cx="16"
            cy="64"
            r="6.438"
            transform="rotate(270 64 64)"
          />
          <circle
            cx="16"
            cy="64"
            r="5.375"
            transform="rotate(315 64 64)"
          />
          <animatetransform
            attributeName="transform"
            calcMode="discrete"
            dur="700ms"
            repeatCount="indefinite"
            type="rotate"
            values="0 64 64;315 64 64;270 64 64;225 64 64;180 64 64;135 64 64;90 64 64;45 64 64"
          />
        </g>
      </svg>
    </span>
  </button>
</DocumentFragment>
`;
